### ISSUE TITLE
Private DNS on Android 9 Blocked

### ISSUE THREAD

**Issue by sphinxir** [ukanth/afwall#985]
my device redmi note 6 pro, miui 10 and android 9, Root with magisk
i set private dns on my device
when afwall is enable dns request doesn't work
when afwall is disable dns request work
plz help

**Comment by CHEF-KOCH** [comment]
We can't help without any log, so please provide a log file. The Wiki also explains how to change your DNS via AFWall+ custom script, which should (in most cases) work.

Did you follow the [instructions shown over here](https://blog.cloudflare.com/enable-private-dns-with-1-1-1-1-on-android-9-pie/)?

**Comment by tusch001** [comment]
On LineageOS 16 (Android 9) private DNS uses tcp protocol / port 853. This is not supported by the app as far as I can read the protocols.
A temporary solution for me is to allow this in a custom script. Most likely it will work on Stock Android 9 too. It's worth a try:

In settings - custom script for start - add an entry to your new file; make sure there is a dot and a space before your path:
. /path/to/your/Script.sh


Add the following two lines (make sure you have the right coding for new line for linux):

IPTABLES=/system/bin/iptables
$IPTABLES -I "afwall" -p tcp --dport 853 -j ACCEPT


This will allow all packets for private DNS.

Hope it helps.

@CHEF-KOCH
Any chance to support this in the final version of 3.2.0?

Thanks in advance.

**Comment by CHEF-KOCH** [comment]
* [DNS over TLS support in Android P Developer Preview](https://security.googleblog.com/2018/04/dns-over-tls-support-in-android-p.html)
* [RFC 7858](https://tools.ietf.org/html/rfc7858)
* `isPrivateDnsActive` [API](https://developer.android.com/reference/android/net/LinkProperties.html#isPrivateDnsActive()
* [Community maintained list of compatible servers](https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Test+Servers).

It's correct we currently do not support the default RFC standards [TCP over 853](https://tools.ietf.org/html/rfc7858#section-3.1) which is mandatory for DoT. There should be an app check if the system own DNS or the new private DNS (Android 9+) is in use. Opening both, port 53 and 853 is dangerous because some apps might perform their own DNS queries. The private dns check is not (yet) implemented.


I'm btw not the owner of this repository, but I help whenver I have some free time. So @ukanth has to look into it. I can't give out any information if this will be addreses in the next stable final version or not, but I don't think so this is not an easy integration because several things must be checked first, as said some apps performing their own DNS quires and if DNS 53 is open (enabled) this might result in some unwanted traffic.

**Comment by tusch001** [comment]
@CHEF-KOCH 

> Opening both, port 53 and 853 is dangerous
Thanks, learned something. I thought that wouldn't matter because I'm whitelisting apps with web access anyway and all others apps wouldn't have access (wrong assumption). So I have a choice between trusting apps on my device or not using private dns

**Comment by markd89** [comment]
What about opening 853 only to the destination IP of the private DNS server? Then the apps on your phone only get to go to one place on port 863.

**Comment by tusch001** [comment]
@markd89
I'm playing around with this in my private custom script.
The problem can not only be solved by allowing access to the name server on port 853. 

Because if port 53 is generally blocked for security reasons when private dns is enabled (as @CHEF-KOCH mentioned it should be) then how can whatever app / program delivers private dns functionality resolve the entry in android setting for private dns? Because - at lineageos16 at least - I can't put an ip address in this field, only a domain name. Without resolving that domain name private dns is not working after reboot.

And as soon as data saving is repeatedly activated / deactivated things are getting complicated. I don't really want to mess with other iptables chains that get changed by the os.

If I would know which package name the app has that provides private dns I could open port 53 only for this app (and the 2-3 others that might need it). I don't know and a blocked port 53 doesn't show on the logs.

Even if I find out this will always be an individual custom script (because of ip address input). Not a one size fits all solution.

**Comment by CHEF-KOCH** [comment]
I had to edit my previous comment multiple times, because my connection timed out (welcome to LTE in Swiss _yay!_). 

### How does Android's "private DNS" really works?

There are some drawbacks which are important to mention.
- The biggest weakness with "private DNS" is that SNI (server name indication) is not encrypted. This means the that you reveal your SNI (server name indication) during the initial TLS negotiation between your device and a specific hostname on a server, to address this there are only handsful of servers which actually supporting encrypted SNI, one of them is [Cloudflare](https://tools.ietf.org/html/draft-ietf-tls-sni-encryption-03).
-  Android DNS field does not accept an IP address, this is by design. Instead, you MUST put in a hostname. Google requires a hostname for this field because of how mobile carriers are adapting to a dual-stack world in which IPv4 and [IPv6](https://www.internetsociety.org/resources/doc/2017/state-of-ipv6-deployment-2017/) coexist. Using IPv6 is not more insecure or fringerprintable than IPv4, I debunked the myth in my fingerprinting project (sr, no link I do not like to advertise my projects here).
- Checking if everything works accordingly can be done via `dig A +short hostname` in your Terminal or via Termux app.
- The data saving option under Android does not affect the private DNS feature, not in LOS nor in any AOSP. If there is something happening which falls, back to the unencrypted DNS then it's maybe a bug in LOS (which I can't confirm, I just tested it). It should not alter iptables.
- There was an app in e.g. F-Droid, (sr I forgot the name .. maybe something with "shark" or so) which can open or list the used API's within an app. To e.g. reveal which app e.g. uses ads services, I guess this app could be used to reveal which installed app has the "isPrivateDnsActive" API which is required for private DNS (other apps simply don't have this API). Another way is to install AdGuard (the trial is not limited in it's logging function) and check the logs after you enabled the proxy mode (see their documentation how you do that). Whenever an app bypasses the proxy this basically means it also bypasses the OS DNS settings.
- Setting Android's "automatic" mode provides a fallback to unencrypted DNS in case Google is not reachable. 
- The option "Private DNS provider hostname" option, should NOT use any fallback this also means n e.g. a mobile connection you are sometimes forced to re-login into the network (that's due to an bug) because to check the network, Android "pings" a Google URL to verify if the network is "okay". I've seen several people saying that Android leaves port 53 open anyway, the specification how this exactly works is given [here](https://blog.apnic.net/2016/06/09/lets-talk-ipv6-dns64-dnssec/).
- LineageOS implementation is identical to the AOSP, only thing I can see is that the pre-choosen option is different (I guess because people have "anti Google concerns"). Besides this, there is no difference.

> What about opening 853 only to the destination IP of the private DNS server? 

This is already been discussed in another ticket, such function would require that the logging mechanism gets an "upgrade" to manually allow (whitelist) the destination IP in order to prevent leaks because some apps like Google Play Services using other (_I call them "sub-apps_") like Google Backup Transport (listed [if installed] under the same "app name") which causes _unknown_ data traffic. The Android Overview only shows the "main app" caused xyz byte/kb/mb traffic. Not all server using the RFC default port (853) some even use 443, so this is not easy to address without opening the ports which are usually "causing" the traffic, 

Theoretically it should not matter, but e.g. WhatsApp checks if 8.8.8.8 is in use, if that's not the case it tries to "overwrite" it (you see this in the Android network logs whenever you block/unblock WhatsApp only), there are (of course) other app examples trying the same. 

As said this is not really an easy "problem". There are a lot of things to consider.

**Comment by tusch001** [comment]
Thanks for all the background info.
I got captive portal login working by allowing backgroud data for it all the time and my own captive portal check (IP is fixed and always allowed).
But as you wrote: not easy. 
My knowledge is too limited so I decided that my system most likely ends up less secure than without using private dns.

Thanks again. I learned a lot.

**Comment by markd89** [comment]
@tusch001 

Could you please post your script so that I (and others) can learn?

Thanks!

**Comment by h1z1** [comment]
The only way to "fix" this is remove it entirely.  

DNS over TLS is an abomination.  Still can't believe it ever got approved.

**Comment by Ghost091180** [comment]
You can use Iptables to do the whole stuff for you and you DONT have to open any INPUT Port. 

Here is my current configuration 
LineageOS16 (crDroid 5.7) Iptables v1.6.1
Magisk v20.1

Create a file at /sbin/.magisk/img/.core/service.d/

Make the file readable with chmod 755

Put in following lines :

#!/system/bin/sh
export PATH=/system/bin

IPTABLES=/system/bin/iptables
IP6TABLES=/system/bin/ip6tables

$IPTABLES -w -F
$IPTABLES -w -t nat -F
$IPTABLES -t nat -N afwall
log  i -t afwall "Flushing Iptables"

$IPTABLES -w -P INPUT DROP
$IPTABLES -w -P FORWARD DROP
$IPTABLES -w -P OUTPUT DROP

$IP6TABLES -w -P INPUT DROP
$IP6TABLES -w -P FORWARD DROP
$IP6TABLES -w -P OUTPUT DROP
log -p i -t afwall "System IPv4/IPv6 policy set to DROP"

$IPTABLES -A INPUT -i lo -p ALL -j ACCEPT
$IPTABLES -A "afwall" -o lo -p ALL -j ACCEPT
$IPTABLES -A INPUT -p ALL -m state --state RELATED,ESTABLISHED -j ACCEPT
$IPTABLES -A INPUT -p ALL -s 192.168.178.1 -j ACCEPT
$IPTABLES -A INPUT -p ALL -j REJECT
log -p i -t afwall "Some IPv4 INPUT Rules set to ACCEPT all oth REJECT"

$IPTABLES -t nat -A OUTPUT -p tcp --dport 853 -j DNAT --to-destination 80.241.218.68:853
$IPTABLES -t nat -A OUTPUT -p udp --dport 853 -j DNAT --to-destination 80.241.218.68:853

$IPTABLES -t nat -A OUTPUT ! -s 192.168.178.0/24 -p tcp --dport 53 -j DNAT --to-destination 80.241.218.68:53
$IPTABLES -t nat -A OUTPUT ! -s 192.168.178.0/24 -p udp --dport 53 -j DNAT --to-destination 80.241.218.68:53
log -p i -t afwall "Set IPv4 DNS Redirection to fdns1.dismail.de"

Where :
192.168.178.1 is the IP of MY Home Network Internet GATEWAY. Change it to yours.

80.241.218.68 is the IP of MY DNS Server (fdns1.dismail.de) Change it to the DNS server IP you want to use.

Save. Restart. Have fun. ;)

IP tables will redirect all local traffic on port 53 and 853 to the pointed destinations and opens a port for related incoming Input and JUST for this.

**Comment by sphinxir** [comment]
> On LineageOS 16 (Android 9) private DNS uses tcp protocol / port 853. This is not supported by the app as far as I can read the protocols.
> A temporary solution for me is to allow this in a custom script. Most likely it will work on Stock Android 9 too. It's worth a try:
> 
> In settings - custom script for start - add an entry to your new file; make sure there is a dot and a space before your path:
> . /path/to/your/Script.sh
> 
> 
> Add the following two lines (make sure you have the right coding for new line for linux):
> 
> IPTABLES=/system/bin/iptables
> $IPTABLES -I "afwall" -p tcp --dport 853 -j ACCEPT
> 
> 
> This will allow all packets for private DNS.
> 
> Hope it helps.
> 
> @CHEF-KOCH
> Any chance to support this in the final version of 3.2.0?
> 
> Thanks in advance.

Thanks for answer
I set custom script and set dns private on my phone.
My problem not solved

**Comment by ukanth** [comment]
Duplicate of #1176

**Comment by t0ma5** [comment]
still no solution for this in 2025 :(
